util/topsql/reporter/subscriber.go

```diff
@@ -0,0 +1,291 @@
+		clientRegistry: clientRegistry,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPublisher{}
+
+// Subscribe registers clients to the reporter and redirects data received from reporter
+// to subscribers associated with those clients.
+func (t *TopSQLPublisher) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	sc := newSubClient(stream, t.decodePlan)
+
+	t.clientRegistry.register(sc)
+	sc.run()
+
+	return nil
+}
+
+type subClient struct {
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+	isDown     *atomic.Bool
+
+	decodePlan planBinaryDecodeFunc
+}
+
+func newSubClient(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+	decodePlan planBinaryDecodeFunc,
+) *subClient {
+	return &subClient{
+		stream:     stream,
+		sendTaskCh: make(chan sendTask),
+		isDown:     atomic.NewBool(false),
+
+		decodePlan: decodePlan,
+	}
+}
+
+func (s *subClient) run() {
+	defer s.isDown.Store(true)
+
+	for task := range s.sendTaskCh {
+		ctx, cancel := context.WithTimeout(context.Background(), task.timeout)
+		start := time.Now()
+
+		var err error
+		doneCh := make(chan struct{})
+		go func(task sendTask) {
+			util.WithRecovery(func() {
+				defer func() {
+					doneCh <- struct{}{}
+				}()
+				err = s.doSend(ctx, task.data)
+				if err != nil {
+					reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+				} else {
+					reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+				}
+			}, nil)
+		}(task)
+
+		select {
+		case <-doneCh:
+			cancel()
+			if err != nil {
+				logutil.BgLogger().Warn(
+					"[top-sql] client failed to send data to subscriber",
+					zap.Error(err),
+				)
+				return
+			}
+		case <-ctx.Done():
+			cancel()
+			logutil.BgLogger().Warn(
+				"[top-sql] client failed to send data to subscriber due to timeout",
+				zap.Duration("timeout", task.timeout),
+			)
+			return
+		}
+	}
+}
+
+func (s *subClient) doSend(ctx context.Context, data reportData) error {
+	if err := s.sendCPUTime(ctx, data.collectedData); err != nil {
+		return err
+	}
+	if err := s.sendSQLMeta(ctx, data.normalizedSQLMap); err != nil {
+		return err
+	}
+	return s.sendPlanMeta(ctx, data.normalizedPlanMap)
+}
+
+func (s *subClient) sendCPUTime(ctx context.Context, data []*dataPoints) (err error) {
+	start := time.Now()
+	sentCount := 0
+	defer func() {
+		topSQLReportRecordCounterHistogram.Observe(float64(sentCount))
+		if err != nil {
+			reportRecordDurationFailedHistogram.Observe(time.Since(start).Seconds())
+		} else {
+			reportRecordDurationSuccHistogram.Observe(time.Since(start).Seconds())
+		}
+	}()
+
+	r := &tipb.TopSQLSubResponse{}
+	record := &tipb.CPUTimeRecord{}
+	r.RespOneof = &tipb.TopSQLSubResponse_Record{Record: record}
+
+	for i := range data {
+		point := data[i]
+		record.SqlDigest = point.SQLDigest
+		record.PlanDigest = point.PlanDigest
+		record.RecordListCpuTimeMs = point.CPUTimeMsList
+		record.RecordListTimestampSec = point.TimestampList
```

Comment:
Please find a way to avoid repeating this in each sink. Then, adding a new sink don't need to have knowledge about exact metrics. Adding new metrics don't need to have knowledge about different sinks as well.